README.md
# ImportExportTools NG modPEC

## Note for non-Italian users 

This is a MOD of a well-known Thunderbird Addon (ImportExportTools NG -
for more details see `README_orig.md`), specifically intended to add
functionalities to backup  [Italian certified email
messages](https://en.wikipedia.org/wiki/Certified_email) and preserve
their legal value as digitally signed documents.

If you are not using Italian Certified Email ("PEC") nor you need to
create a hash list of your messages saved in EML format, you probably do
not need this MOD at all.

For the above reason, the remainder of this README file is in Italian
language, and the active localizations are only Italian and English-US.

## Nota importante

Questo MOD è un fork del nuovo addon "ImportExportTools NG", compatibile
con le versioni di Thunderbird successive alla 60.

Il MOD del vecchio addon, compatibile con le precedenti versioni di
Thunderbird, è disponibile
[qui](https://github.com/alpianon/importexporttools-modpec).

## A cosa serve?  

Questo MOD aggiunge all'addon importexporttools delle funzionalità che
aiutano l'utente a fare il backup dei messaggi di PEC in modo da poterne
provare in futuro l'autenticità, nel rispetto della privacy e della
riservatezza della corrispondenza.  L'idea è nata da una conversazione
tra legali ([l'autore di questo MOD](http://pianon.eu/) e [Carlo
Piana](https://www.piana.eu/), che ha suggerito la procedura che viene
qui implementata) nella mailing list di [ARRAY](https://www.array.eu/)
in merito ai problemi pratici di conservazione delle PEC. Per
approfondire, leggete più avanti il paragrafo sugli aspetti legali.

**ATTENZIONE Questo MOD è ancora work-in-progress, non è ancora stato
testato in produzione in modo adeguato, ed è inteso unicamente come
applicazione da provare in ambienti di test e sviluppo. Si declina
qualsiasi responsabilità per qualsiasi uso in produzione e a fini
professionali.**

*Ferma l'esclusione di garanzia di cui sopra, si può verificare la
correttezza della generazione della lista di hash con lo script
`debug.sh` creato automaticamente dentro alla cartella di backup, come
spiegato meglio più avanti (vale solo per gli ambienti linux).*

## Installazione

Se avete ImportExportTools NG (versione originale) installato, è
opportuno disinstallarlo, prima di installare questo MOD.

Per installare ImportExportTools NG "moddato" è sufficiente scaricare,
dalla cartella "xpi", il file con estensione `.xpi` relativo all'ultima
versione disponibile, e poi installarlo dalla scheda "Gestione
componenti aggiuntivi" di Thunderbird, cliccando sull'icona delle
impostazioni (quella a forma di ruota dentata, per capirci), e
selezionando la voce "Installa componente aggiuntivo da file".

### NOTA IMPORTANTE PER GLI UTENTI LINUX

Come ulteriore funzionalità aggiunta, questo MOD consente di scegliere,
nelle opzioni di ImportExportTools, di **tagliare il nome dei file
esportati a 140 caratteri** invece che "solo" a 256 caratteri; questo
perchè, se nella generalità dei filesystem il limite è 256 caratteri,
quando si usa Ecryptfs (tipicamente, se si sceglie di avere la cartella
home cifrata in certe distribuzioni Linux) il limite è di fatto 140
caratteri, come spiegato meglio
[qui](https://unix.stackexchange.com/questions/32795/what-is-the-maximum-allowed-filename-and-folder-size-with-ecryptfs)).
Con questa opzione, è possibile salvare i messaggi da importexporttools
anche in una cartella cifrata con Ecryptfs senza incorrere nell'errore
`NS_ERROR_FILE_NAME_TOO_LONG` se l'oggetto dell'email (dal quale
importexporttools genera il nome del file) è troppo lungo.

## Come si usa?

Alle tradizionali opzioni di menu di importexporttools, questo MOD
aggiunge un'opzione per salvare i messaggi *"come file EML con hash list
e verifica SMIME (Backup PEC)"* (nella finestra "cerca ed esporta"
l'opzione è indicata come *"EML+hash list+verifica SMIME"* per motivi di
spazio ma la funzionalità è la stessa).

Selezionando tale opzione, verrà creata una cartella (con il nome della
cartella di Thunderbird e la data e ora di esportazione, se viene
esportata un'intera cartella; con il prefisso "Backup_PEC" e la data e
ora di esportazione, se vengono esportati solo i messaggi selezionati o
cercati).

All'interno di questa cartella, alla fine dell'esportazione, troverete
una sottocartella "messaggi" (contenente i messaggi di PEC esportati in
formato EML) e due file, `lista_hash.txt` e
`lista_hash_con_nomi_file.txt`: il contenuto del primo andrà inserito in
apposita dichiarazione da firmare digitalmente e marcare temporalmente,
il secondo andrà conservato insieme al backup come riferimento futuro
(vd. più avanti il paragrafo relativo agli aspetti legali).

  * **Importante**: Il contenuto del file `lista_hash_con_nomi_file.txt`
    è volutamente identico all'output di un comando `sha256sum * | sort`
    lanciato nella sottocartella `messaggi` dentro alla cartella di
    backup, in modo da consentire di fare una verifica incrociata sul
    corretto funzionamento dell'addon.
  * Questa verifica può essere effettuata in modo "automatico" facendo
    girare lo script debug.sh che viene generato dentro alla cartella
    di backup (`cd percorso/cartella/backup/; ./debug.sh`)

Durante l'esportazione dei messaggi di PEC, viene anche effettuata una
**verifica della firma digitale del provider di PEC (Aruba, Legalmail,
ecc.)** all'interno di ciascun messaggio: se la firma di alcuni messaggi
non è valida o è scaduta, alla fine dell'esportazione viene visualizzato
un messaggio di errore, con indicazione di guardare, per i dettagli, il
file `errori_verifica_PEC.txt` nella cartella di esportazione.

  * **Attenzione**: nel caso vengano rilevati degli errori su alcuni
    messaggi esportati, la lista degli hash viene comunque generata; ma
    l'operazione di firma digitale e marcatura temporale su tale lista
    non ha alcun valore/senso dal punto di vista legale, o meglio non ne
    ha relativamente ai messaggi sui quali sono stati rilevati degli
    errori (vd. sempre più avanti il paragrafo relativo agli aspetti
    legali).

## Aspetti legali relativi al backup dei messaggi di PEC

### Perché, nel caso dei messaggi di PEC, un backup "tradizionale" non basta?

La normativa sui documenti informatici prevede che un documento firmato
con un certificato scaduto non è un documento valido, equivale a un
documento non firmato. Nel momento in cui il certificato viene
utilizzato per la firma, il documento firmato è valido. Chiunque può
osservare, da quel momento in poi, che il documento è stato firmato con
un certificato valido, usando uno degli strumenti informatici (alcuni
anche _open source_) messi a disposizione per la verifica. Tuttavia i
certificati **scadono**. Pertanto, dal giorno in cui il certificato con
cui il documento è firmato scade, non è più possibile sapere con
certezza se al momento della firma il certificato fosse valido. Pertanto
viene a cessare la funzione di prova opponibile ai terzi
dell'autenticità. Per conservare tale efficacia probatoria, occorre uno
strumento che dia certezza temporale del fatto che a un dato momento,
antecedente alla scadenza del certificato, il documento fosse già stato
firmato.

Questo procedimento è appunto la conservazione. Tenere una o più copie
fisiche del file, infatti, non garantisce il **tempo** in cui la firma
contenuta nel file fosse stata apposta. Solo con un riferimento
riconosciuto dalla legge come opponibile ai terzi, ciò è possibile.
Apponendo la marca temporale, un terzo certifica che quel documento
esisteva già al momento di marcatura. Dunque fornisce la prova
necessaria a conservare la validità giuridica al documento informatico
così realizzato.

### La marca temporale su un archivio di messaggi di PEC: problemi di privacy

L'apposizione di una marca temporale a tutto l'archivio di messaggi PEC
porterebbe ‒ nel caso si dovesse provare la data di firma di un
determinato messaggio ‒ alla necessità di produrre l'intera porzione di
archivio. Non sarebbe infatti possibile validare la marca temporale
senza avere l'esatta sequenza di bit a cui è stata apposta.
L'alternativa sarebbe apporre la marcatura temporale a ogni singolo
messaggio, cosa che ha un costo elevato e una praticità molto scarsa.

Ulteriore alternativa è appunto usare un **elenco di hash**, in modo che
la marca temporale sia univocamente collegata a una serie di messaggi,
ma non sia necessario produrre tutti i messaggi, al fine di provare che
quel messaggio in quella esatta e inalterata sequenza di bit, è stato
creato prima della scadenza del relativo certificato.

D'altronde, ogni firma, inclusa la marca temporale, consiste nella
cifratura di un hash del documento (non del documento) con la chiave
privata. La firma di un documento contenente gli hash è
computazionalmente equivalente.

### Una soluzione per fare il backup dei messaggi di PEC nel rispetto della privacy

Nella soluzione qui proposta, usiamo due liste. La
`lista_hash_con_nomi_file` contiene l'elenco completo con i nomi di
file. La `lista_hash_noname.txt` è l'elenco delle stesse impronte hash
senza i nomi di file, che può essere così inserito in un documento senza
nemmeno rivelare l'esistenza di un determinato messaggio, visto che
anche il solo titolo del messaggio, pur accorciato, potrebbe rivelare
informazioni.

Qualora occorresse produrre in giudizio o in altra sede il messaggio
relativo, a certificato scaduto, sarebbe possibile creare un **pacchetto
di distribuzione** mediante l'inserimento in un contenitore (un file
zip, ad esempio) del messaggio originale e il documento con gli hash,
firmato e marcato temporalmente. Usando `lista_hash_con_nomi_file` come
tabella di corrispondenza è possibile risalire univocamente e
agevolmente da un file al documento contenente gli hash e viceversa. La
perdita del file `lista_hash_con_nomi_file` non sarebbe comunque grave,
potendolo in ogni momento ricostruire (non applicando un _salt_, ogni
documento ha uno e un solo hash SHA256).


CHANGELOG.md
# ImportExportTools NG modPEC Changelog

## Versions

4.0.4-b1 - November 30, 2019
- ported modPEC for old ImportExportTools to ImportExportTools NG
- everything seems to work, but it needs some more testing

## Original NG Extension

4.0.4 Maintenance  Release - September 16, 2019 
- Fixed text formatter signature change
- Fixed handling of import filenames > 55 characters

4.0.3 Maintenance  Release - September 12, 2019 
- Fixed EML import issue
- Fixed HTML layout for text attachments

4.0.2 Maintenance  Release - September 10, 2019 
- Fixed plaintext/HTML attachment export
  String.trim test failing - String generics deprecated
  Fixes #13
- Remove remaining Components.*
- Fix HTML export format for embedded images
  Change file pointer fixup 
  Move image into fieldset

4.0.1 Maintenance  Release - August 15, 2019 
 - Fixed warning for deprecated menu overlay task popup
   Fixes #9 
 - Use createXULElement for TB68+
   Fixes #8 , TB69 Beta compatibility

4.0.0  Initial NG Release - July 19, 2019 
 - Updated for TB68
 - New name and icon
 - Updated formatting
 - Options dialog tweaks, icon

## Original Extension:

3.3.2  Update for TB60 - January 27, 2019
 - Last version update by "Kaosmos"


src/chrome/content/mboximport/mboximportOptions.js
/*
  ImportExportTools NG modPEC id a fork of ImportExportTools NG with added
  functionalities to export and verify SMIME messages and generate a hash
  list of the exported messages. It is intended to be used to backup
  Italian Certified Email (PEC) messages.
  The MOD author:
    Copyright (C) 2019 : Alberto Pianon

	ImportExportTools NG is a derivative extension for Thunderbird 60+
	providing import and export tools for messages and folders.
	The derivative extension authors:
		Copyright (C) 2019 : Christopher Leidigh, The Thunderbird Team

	The original extension & derivatives, ImportExportTools, by Paolo "Kaosmos",
	is covered by the GPLv3 open-source license (see LICENSE file).
		Copyright (C) 2007 : Paolo "Kaosmos"

	ImportExportTools NG is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

// cleidigh - reformat, services, globals, dialog changes

/* global IETprefs, IETgetComplexPref, IETsetComplexPref, IETpickFile */

var { Services } = ChromeUtils.import('resource://gre/modules/Services.jsm');

function IETsetCharsetPopup(charsetPref) {
    const versionChecker = Services.vc;
    const currentVersion = Services.appinfo.platformVersion;


    var charsetPopup = document.getElementById("charset-list-popup");
    var charsetList = IETprefs.getCharPref("extensions.importexporttoolsngmodpec.export.charset_list");
    var charsetItems = charsetList.split(",");
    var menuitem;

    for (var i = 0; i < charsetItems.length; i++) {
        if (versionChecker.compare(currentVersion, "68") >= 0) {
            // replacement for createElement post TB60
            menuitem = document.createXULElement("menuitem");
        } else {
            menuitem = document.createElement("menuitem");
        }

        menuitem.setAttribute("label", charsetItems[i]);
        menuitem.setAttribute("value", charsetItems[i]);
        charsetPopup.appendChild(menuitem);
        if (charsetItems[i] === charsetPref)
            document.getElementById("charset-list").selectedItem = menuitem;
    }
}

function initMboxImportPanel() {

    const versionChecker = Services.vc;
    const currentVersion = Services.appinfo.platformVersion;

    // cleidigh - TB68 groupbox needs hbox/label
    if (versionChecker.compare(currentVersion, "61") >= 0) {
        var captions = document.querySelectorAll("caption");
        for (let i = 0; i < captions.length; i++) {
            captions[i].style.display = "none";
        }
    } else {
        var groupboxtitles = document.querySelectorAll(".groupbox-title");
        for (let i = 0; i < groupboxtitles.length; i++) {
            groupboxtitles[i].style.display = "none";
        }
    }

    IETsetCharsetPopup("");

    document.getElementById("MBoverwrite").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.overwrite");
    document.getElementById("MBasciiname").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.filenames_toascii");
    document.getElementById("MBconfrimimport").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.confirm.before_mbox_import");
    document.getElementById("MBhtmlasdisplayed").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.HTML_as_displayed");
    document.getElementById("MBcliptextplain").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.clipboard.always_just_text");
    document.getElementById("MBsubmaxlen").value = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.subject.max_length");
    document.getElementById("MBauthmaxlen").value = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.author.max_length");
    document.getElementById("MBrecmaxlen").value = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.recipients.max_length");
    document.getElementById("setTimestamp").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.set_filetime");
    document.getElementById("addtimeCheckbox").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.filenames_addtime");
    document.getElementById("buildMSF").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.import.build_mbox_index");
    document.getElementById("addNumber").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.import.name_add_number");

    if (IETprefs.getIntPref("extensions.importexporttoolsngmodpec.exportEML.filename_format") === 2)
        document.getElementById("customizeFilenames").checked = true;
    else
        document.getElementById("customizeFilenames").checked = false;

    if (IETprefs.getPrefType("extensions.importexporttoolsngmodpec.exportMBOX.dir") > 0)
        document.getElementById("export_mbox_dir").value = IETgetComplexPref("extensions.importexporttoolsngmodpec.exportMBOX.dir");
    if (IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.exportMBOX.use_dir")) {
        document.getElementById("use_export_mbox_dir").checked = true;
        document.getElementById("export_mbox_dir").removeAttribute("disabled");
        document.getElementById("export_mbox_dir").nextSibling.removeAttribute("disabled");
    } else {
        document.getElementById("use_export_mbox_dir").checked = false;
        document.getElementById("export_mbox_dir").setAttribute("disabled", "true");
        document.getElementById("export_mbox_dir").nextSibling.setAttribute("disabled", "true");
    }

    if (IETprefs.getPrefType("extensions.importexporttoolsngmodpec.exportEML.dir") > 0)
        document.getElementById("export_eml_dir").value = IETgetComplexPref("extensions.importexporttoolsngmodpec.exportEML.dir");
    if (IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.exportEML.use_dir")) {
        document.getElementById("use_export_eml_dir").checked = true;
        document.getElementById("export_eml_dir").removeAttribute("disabled");
        document.getElementById("export_eml_dir").nextSibling.removeAttribute("disabled");
    } else {
        document.getElementById("use_export_eml_dir").checked = false;
        document.getElementById("export_eml_dir").setAttribute("disabled", "true");
        document.getElementById("export_eml_dir").nextSibling.setAttribute("disabled", "true");
    }

    if (IETprefs.getPrefType("extensions.importexporttoolsngmodpec.exportMSG.dir") > 0)
        document.getElementById("export_msgs_dir").value = IETgetComplexPref("extensions.importexporttoolsngmodpec.exportMSG.dir");
    if (IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.exportMSG.use_dir")) {
        document.getElementById("use_export_msgs_dir").checked = true;
        document.getElementById("export_msgs_dir").removeAttribute("disabled");
        document.getElementById("export_msgs_dir").nextSibling.removeAttribute("disabled");
    } else {
        document.getElementById("use_export_msgs_dir").checked = false;
        document.getElementById("export_msgs_dir").setAttribute("disabled", "true");
        document.getElementById("export_msgs_dir").nextSibling.setAttribute("disabled", "true");
    }

    if (IETprefs.getPrefType("extensions.importexporttoolsngmodpec.export.filename_pattern") > 0) {
        var pattern = IETprefs.getCharPref("extensions.importexporttoolsngmodpec.export.filename_pattern");
        var patternParts = pattern.split("-");

        for (var i = 0; i < 3; i++) {
            var list = document.getElementById(`part${i + 1}`);
            var popup = document.getElementById(`part${i + 1}-popup-list`);

            switch (patternParts[i]) {
                case "%d":
                    list.selectedItem = popup.childNodes[1];
                    break;
                case "%k":
                    list.selectedItem = popup.childNodes[2];
                    break;
                case "%n":
                    list.selectedItem = popup.childNodes[3];
                    break;
                case "%a":
                    list.selectedItem = popup.childNodes[4];
                    break;
                case "%r":
                    list.selectedItem = popup.childNodes[5];
                    break;
                case "%e":
                    list.selectedItem = popup.childNodes[6];
                    break;
                default:
                    list.selectedItem = popup.childNodes[0];
            }
        }
    }

    document.getElementById("addPrefix").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.filename_add_prefix");
    try {
        document.getElementById("prefixText").value = IETgetComplexPref("extensions.importexporttoolsngmodpec.export.filename_prefix");
    } catch (e) {}

    document.getElementById("cutSub").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.cut_subject");
    document.getElementById("cutFN").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.cut_filename");
    customNamesCheck(document.getElementById("customizeFilenames"));

    var charset = "";
    var textCharset = "";
    var csvSep = "";

    try {
        charset = IETprefs.getCharPref("extensions.importexporttoolsngmodpec.export.filename_charset");
        textCharset = IETprefs.getCharPref("extensions.importexporttoolsngmodpec.export.text_plain_charset");
        csvSep = IETprefs.getCharPref("extensions.importexporttoolsngmodpec.csv_separator");
    } catch (e) {
        charset = "";
        textCharset = "";
        csvSep = "";
    }

    IETsetCharsetPopup(textCharset);
    document.getElementById("filenameCharset").value = charset;
    document.getElementById("csvSep").value = csvSep;

    document.getElementById("skipMsg").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.skip_existing_msg");
    if (IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.export.use_container_folder")) {
        document.getElementById("indexSetting").selectedIndex = 0;
        document.getElementById("skipMsg").disabled = true;
    } else {
        document.getElementById("indexSetting").selectedIndex = 1;
    }

    //modPEC 
 if (IETprefs.getIntPref("extensions.importexporttoolsngmodpec.export.cut_filename_at") == 256)
        document.getElementById("cutFNat").selectedIndex = 0;
    else
      document.getElementById("cutFNat").selectedIndex = 1;
    //modPEC

    // Backup section
    var freq = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.frequency");

    switch (freq) {
        case 1:
            document.getElementById("frequencyList").selectedIndex = 0;
            document.getElementById("backupEnable").checked = true;
            break;
        case 3:
            document.getElementById("frequencyList").selectedIndex = 1;
            document.getElementById("backupEnable").checked = true;
            break;
        case 7:
            document.getElementById("frequencyList").selectedIndex = 2;
            document.getElementById("backupEnable").checked = true;
            break;
        case 15:
            document.getElementById("frequencyList").selectedIndex = 3;
            document.getElementById("backupEnable").checked = true;
            break;
        case 30:
            document.getElementById("frequencyList").selectedIndex = 4;
            document.getElementById("backupEnable").checked = true;
            break;
        default:
            document.getElementById("backupEnable").checked = false;
            document.getElementById("frequencyList").disabled = true;
    }

    try {
        document.getElementById("backupDir").value = IETgetComplexPref("extensions.importexporttoolsngmodpec.autobackup.dir");
        document.getElementById("backupCustomName").value = IETgetComplexPref("extensions.importexporttoolsngmodpec.autobackup.dir_custom_name");
    } catch (e) {}

    document.getElementById("backupType").selectedIndex = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.type");
    var dir = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.dir_name_type");
    document.getElementById("backupDirName").selectedIndex = dir;
    document.getElementById("backupType").selectedIndex = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.type");
    document.getElementById("saveMode").selectedIndex = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.save_mode");

    var last = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.last") * 1000;
    if (last > 0) {
        var time = new Date(last);
        var localTime = time.toLocaleString();
        document.getElementById("backupLast").value = localTime;
    }
    document.getElementById("modalWin").checked = IETprefs.getBoolPref("extensions.importexporttoolsngmodpec.autobackup.use_modal_dialog");
}

/* function setSaveMode(type) {
	var saveMode = IETprefs.getIntPref("extensions.importexporttoolsngmodpec.autobackup.save_mode");
	if (saveMode == 0 || (saveMode == 2 && type ==0))
		document.getElementById("saveMode").selectedIndex = 0;
	else
		document.getElementById("saveMode").selectedIndex = 1;
}

function toggleType(el) {
	setSaveMode(el.selectedIndex);
}*/

function saveMboxImportPrefs() {
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.overwrite", document.getElementById("MBoverwrite").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.filenames_toascii", document.getElementById("MBasciiname").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.confirm.before_mbox_import", document.getElementById("MBconfrimimport").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.HTML_as_displayed", document.getElementById("MBhtmlasdisplayed").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.clipboard.always_just_text", document.getElementById("MBcliptextplain").checked);
    IETprefs.setIntPref("extensions.importexporttoolsngmodpec.subject.max_length", document.getElementById("MBsubmaxlen").value);
    IETprefs.setIntPref("extensions.importexporttoolsngmodpec.author.max_length", document.getElementById("MBauthmaxlen").value);
    IETprefs.setIntPref("extensions.importexporttoolsngmodpec.recipients.max_length", document.getElementById("MBrecmaxlen").value);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.set_filetime", document.getElementById("setTimestamp").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.filenames_addtime", document.getElementById("addtimeCheckbox").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.import.build_mbox_index", document.getElementById("buildMSF").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.import.name_add_number", document.getElementById("addNumber").checked);

    if (document.getElementById("customizeFilenames").checked)
        IETprefs.setIntPref("extensions.importexporttoolsngmodpec.exportEML.filename_format", 2);
    else
        IETprefs.setIntPref("extensions.importexporttoolsngmodpec.exportEML.filename_format", 0);

    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.exportMBOX.use_dir", document.getElementById("use_export_mbox_dir").checked);
    if (document.getElementById("export_mbox_dir").value !== "")
        IETsetComplexPref("extensions.importexporttoolsngmodpec.exportMBOX.dir", document.getElementById("export_mbox_dir").value);
    else
        IETprefs.deleteBranch("extensions.importexporttoolsngmodpec.exportMBOX.dir");

    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.exportEML.use_dir", document.getElementById("use_export_eml_dir").checked);
    if (document.getElementById("export_eml_dir").value !== "")
        IETsetComplexPref("extensions.importexporttoolsngmodpec.exportEML.dir", document.getElementById("export_eml_dir").value);
    else
        IETprefs.deleteBranch("extensions.importexporttoolsngmodpec.exportEML.dir");

    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.exportMSG.use_dir", document.getElementById("use_export_msgs_dir").checked);
    if (document.getElementById("export_msgs_dir").value !== "")
        IETsetComplexPref("extensions.importexporttoolsngmodpec.exportMSG.dir", document.getElementById("export_msgs_dir").value);
    else
        IETprefs.deleteBranch("extensions.importexporttoolsngmodpec.exportMSG.dir");

    var pattern = "";
    for (let u = 1; u < 4; u++) {
        var val = document.getElementById("part" + u.toString()).selectedItem.value;
        if (u > 1 && val)
            val = "-" + val;
        pattern += val;
    }
    IETprefs.setCharPref("extensions.importexporttoolsngmodpec.export.filename_pattern", pattern);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.filename_add_prefix", document.getElementById("addPrefix").checked);
    // if (document.getElementById("prefixText").value != "")
    IETsetComplexPref("extensions.importexporttoolsngmodpec.export.filename_prefix", document.getElementById("prefixText").value);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.cut_subject", document.getElementById("cutSub").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.cut_filename", document.getElementById("cutFN").checked);
    //modPEC
    if (document.getElementById("cutFNat").selectedIndex === 0)       
       IETprefs.setIntPref("extensions.importexporttoolsngmodpec.export.cut_filename_at", 256);       
     else
       IETprefs.setIntPref("extensions.importexporttoolsngmodpec.export.cut_filename_at", 140);
    //modPEC
    
    IETprefs.setCharPref("extensions.importexporttoolsngmodpec.export.filename_charset", document.getElementById("filenameCharset").value);
    IETprefs.setCharPref("extensions.importexporttoolsngmodpec.export.text_plain_charset", document.getElementById("charset-list").selectedItem.value);
    IETprefs.setCharPref("extensions.importexporttoolsngmodpec.csv_separator", document.getElementById("csvSep").value);

    if (document.getElementById("indexSetting").selectedIndex === 0)
        IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.use_container_folder", true);
    else
        IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.use_container_folder", false);

    // Backup section
    if (!document.getElementById("backupEnable").checked)
        IETprefs.setIntPref("extensions.importexporttoolsngmodpec.autobackup.frequency", 0);
    else
        IETprefs.setIntPref("extensions.importexporttoolsngmodpec.autobackup.frequency", document.getElementById("frequencyList").selectedItem.value);
    if (document.getElementById("backupDir").value)
        IETsetComplexPref("extensions.importexporttoolsngmodpec.autobackup.dir", document.getElementById("backupDir").value);
    else
        IETprefs.deleteBranch("extensions.importexporttoolsngmodpec.autobackup.dir");
    IETprefs.setIntPref("extensions.importexporttoolsngmodpec.autobackup.dir_name_type", document.getElementById("backupDirName").selectedIndex);
    if (document.getElementById("backupCustomName").value)
        IETsetComplexPref("extensions.importexporttoolsngmodpec.autobackup.dir_custom_name", document.getElementById("backupCustomName").value);
    else
        IETprefs.deleteBranch("extensions.importexporttoolsngmodpec.autobackup.dir_custom_name");
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.export.skip_existing_msg", document.getElementById("skipMsg").checked);
    IETprefs.setBoolPref("extensions.importexporttoolsngmodpec.autobackup.use_modal_dialog", document.getElementById("modalWin").checked);
    IETprefs.setIntPref("extensions.importexporttoolsngmodpec.autobackup.type", document.getElementById("backupType").selectedIndex);
    IETprefs.setIntPref("extensions.importexporttoolsngmodpec.autobackup.save_mode", document.getElementById("saveMode").selectedIndex);
}

function customNamesCheck(el) {
    if (!el.checked) {
        document.getElementById("addtimeCheckbox").setAttribute("disabled", "true");
        document.getElementById("part1").setAttribute("disabled", "true");
        document.getElementById("part2").setAttribute("disabled", "true");
        document.getElementById("part3").setAttribute("disabled", "true");
        document.getElementById("addPrefix").setAttribute("disabled", "true");
        document.getElementById("prefixText").setAttribute("disabled", "true");
    } else {
        document.getElementById("addtimeCheckbox").removeAttribute("disabled");
        document.getElementById("part1").removeAttribute("disabled");
        document.getElementById("part2").removeAttribute("disabled");
        document.getElementById("part3").removeAttribute("disabled");
        document.getElementById("addPrefix").removeAttribute("disabled");
        document.getElementById("prefixText").removeAttribute("disabled");
    }
}

function toggleDirCheck(el) {
    if (!el.checked) {
        el.nextSibling.setAttribute("disabled", "true");
        el.nextSibling.nextSibling.setAttribute("disabled", "true");
    } else {
        el.nextSibling.removeAttribute("disabled");
        el.nextSibling.nextSibling.removeAttribute("disabled");
    }
}

function toggleBackup(el) {
    document.getElementById("frequencyList").disabled = !el.checked;
}

function toggleSkipMsg(el) {
    document.getElementById("skipMsg").disabled = (el.selectedIndex === 0);
}

//modPEC
function toggleCutFNat(el) {
	document.getElementById("cutFNat").disabled = ! el.checked;
}
//modPEC

function pickFile(el) {
    IETpickFile(el);
}

document.addEventListener("dialogaccept", function(event) {
    saveMboxImportPrefs();
});

window.addEventListener("load", function(event) {
    initMboxImportPanel();
});
